core/drivers/rstctrl/stm32mp1_rstctrl.c

```diff
@@ -0,0 +1,173 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (c) 2021-2022, Linaro Limited
+ * Copyright (c) 2018-2022, STMicroelectronics
+ */
+
+#include <drivers/rstctrl.h>
+#include <drivers/stm32mp_dt_bindings.h>
+#include <drivers/stm32mp1_rcc.h>
+#include <io.h>
+#include <keep.h>
+#include <kernel/delay.h>
+#include <kernel/dt.h>
+#include <kernel/dt_driver.h>
+#include <kernel/panic.h>
+#include <mm/core_memprot.h>
+#include <stm32_util.h>
+
+#include "stm32_rstctrl.h"
+
+#define RESET_ID_MASK		GENMASK_32(31, 5)
+#define RESET_ID_SHIFT		U(5)
+#define RESET_BIT_POS_MASK	GENMASK_32(4, 0)
+#define RESET_OFFSET_MAX	U(1024)
+
+static size_t reset_id2reg_offset(unsigned int id)
+{
+	size_t offset = (id & RESET_ID_MASK) >> RESET_ID_SHIFT;
+
+	assert(offset < RESET_OFFSET_MAX);
+	return offset * sizeof(uint32_t);
+}
+
+static uint8_t reset_id2reg_bit_pos(unsigned int reset_id)
+{
+	uint8_t pos = reset_id & RESET_BIT_POS_MASK;
+
+	assert(pos < 32);
+	return pos;
+}
+
+static TEE_Result reset_assert(struct rstctrl *rstctrl, unsigned int to_us)
+{
+	unsigned int id = to_rstline(rstctrl)->id;
+	vaddr_t rcc_base = stm32_rcc_base();
+	uint32_t bit_mask = 0;
+	size_t offset = 0;
+
+#ifdef CFG_STM32MP15
+	switch (id) {
+	case MCU_HOLD_BOOT_R:
+		/*
+		 * The RCC_MP_GCR is a read/write register.
+		 * Assert the MCU HOLD_BOOT means clear the BOOT_MCU bit
+		 */
+		io_clrbits32(rcc_base + RCC_MP_GCR, RCC_MP_GCR_BOOT_MCU);
+
+		return TEE_SUCCESS;
+	case MCU_R:
+		/* MCU reset can only be written */
+		to_us = 0;
+		break;
+	default:
+		break;
+	}
+#endif
+
+	offset = reset_id2reg_offset(id);
+	bit_mask = BIT(reset_id2reg_bit_pos(id));
+
+	io_write32(rcc_base + offset, bit_mask);
+
+	if (to_us) {
+		uint64_t timeout_ref = timeout_init_us(to_us);
+
+		while (!(io_read32(rcc_base + offset) & bit_mask))
+			if (timeout_elapsed(timeout_ref))
+				break;
+
+		if (!(io_read32(rcc_base + offset) & bit_mask))
+			return TEE_ERROR_SECURITY;
```

Comment:
TEE_ERROR_BAD_STATE might be better

Reply:
Indeed this is not a security specific error. `TEE_ERROR_GENERIC` is often used in OP-TEE core as return code on timeout issues, specifically in the other stm32_xxx drivers.